src/nb_type.cpp

```diff
@@ -336,101 +335,234 @@ static void inst_dealloc(PyObject *self) {
     Py_DECREF(tp);
 }
 
+#if defined(NB_FREE_THREADED)
+nb_type_map_per_thread::guard nb_type_lock_c2p_fast(nb_internals *internals_)
+    noexcept {
+    thread_local nb_type_map_per_thread type_c2p_per_thread{*internals_};
+    return type_c2p_per_thread.lock();
+}
+#endif
+
+static void nb_type_c2p_fill_from_slow(nb_internals *internals_,
+                                       void *&fast_value,
+                                       const std::type_info *type) {
+    // Cache miss. Fetch the true value from the slow map.
+    nb_type_map_slow &type_c2p_slow = internals_->type_c2p_slow;
+    nb_type_map_slow::iterator it_slow = type_c2p_slow.find(type);
+    fast_value = (it_slow != type_c2p_slow.end() ? it_slow->second : nullptr);
+
+    // Maintain a linked list to clean up 'type_c2p_fast' when the type
+    // expires (see nb_type_unregister).
+    auto [it_alias, inserted] =
+        internals_->types_in_c2p_fast.try_emplace(type, nullptr);
+    if (!inserted) {
+        // Check whether `type` is an alias we haven't seen yet.
+        bool seen = (type == it_alias->first);
+        nb_alias_seq *prev = nullptr;
+        nb_alias_seq *node = (nb_alias_seq *) it_alias->second;
+        while (node && !seen) {
+            seen |= (type == node->value);
+            prev = node;
+            node = node->next;
+        }
+        if (!seen) {
+            // Got a new one -- add it. NB: use non-Python allocator for these
+            // since we won't free them until internals_cleanup(), which is
+            // after interpreter finalization
+            node = (nb_alias_seq *) malloc(sizeof(nb_alias_seq));
+            check(node, "Could not allocate alias chain entry!");
+            node->value = type;
+            node->next = nullptr;
+            if (prev)
+                prev->next = node;
+            else
+                it_alias.value() = node;
+        }
+    }
+}
 
 type_data *nb_type_c2p(nb_internals *internals_,
-                       const std::type_info *type) {
+                       const std::type_info *type,
+                       bool *has_foreign) {
 #if defined(NB_FREE_THREADED)
-    thread_local nb_type_map_fast type_c2p_fast;
+    auto per_thread_guard = nb_type_lock_c2p_fast(internals_);
+    nb_type_map_fast &type_c2p_fast = *per_thread_guard;
 #else
     nb_type_map_fast &type_c2p_fast = internals_->type_c2p_fast;
 #endif
+    void *const sentinel = (void *) (uintptr_t) 1;
```

Comment:
This sentinel construction is worthy of a comment, it seems more complicated than what we had here before.